### PR TITLE
Fix #1041

### LINKED ISSUE #1041: Client keep-alive support is broken
Hello,

I recently added this http library to server HTTP requests within the ccache compiler cache. Because a compilation requires up four HTTP requests I enabled keep-alive support in the Client.

But we observed a race condition on Linux:

1. Connection is opened and a GET request is issued. Turns out to be a cache miss
2. Compilation starts and takes 90s, meanwhile the server closes the connection
3. A PUT is issued on the now closed connection
4. The ccache process receives SIGPIPE.

Somehow the `is_alive` detection does not work as expected. A rough test case looks like this:

```c++
TEST(KeepAlive, SimpleInterface_Online) {

  const auto host = "127.0.0.1";
  const auto port = 8080;
  const auto resourcePath = "/hi";

  Server svr;

  svr.set_keep_alive_timeout(3);

  svr.Get(resourcePath, [](const httplib::Request &, httplib::Response &res) {
    res.set_content("Hello World!", "text/plain");
  });

  auto a2 = std::async(std::launch::async, [&svr, host, port]{ svr.listen(host, port); });

  std::this_thread::sleep_for(std::chrono::milliseconds(200));

  Client cli(host, port);
  cli.set_keep_alive(true);

  auto result = cli.Get(resourcePath);
  ASSERT_TRUE(result);
  EXPECT_EQ(200, result->status);

  std::this_thread::sleep_for(std::chrono::seconds(5));

  result = cli.Get(resourcePath);
  ASSERT_TRUE(result);
  EXPECT_EQ(200, result->status);

  svr.stop();
  a2.wait();
}
```

Besides the non-working `is_alive` detection the library should set the `MSG_NOSIGNAL` on Linux and `SO_NOSIGPIPE` on Apple platforms to not badly interfere with the library host process.

Thanks,
Gregor

### PR DESCRIPTION
**Pull request by yhirose**

### CODE REVIEW

**Review by tsilia — CHANGES_REQUESTED** [review]
I've checked and the logic I had described in the comment (see below) works fine: client doesn't close socket after each send.
So I'm suggesting you to re-implement `is_socket_alive` like this:
```cpp
inline bool is_socket_alive(socket_t sock) {
	if (detail::select_read(sock, 0, 0) == 0)
		return true;	// assume the socket is still alive
	char buf[1];
	const auto recv_res = recv(sock, &buf[0], sizeof(buf), MSG_PEEK);
	return (recv_res > 0);
}
```
Please note that the code doesn't handle possible errors.

Thanks!

**Review by tsilia — APPROVED** [review]
Works like a charm on my Linux system, thanks!

Though I'd recommend at least handle `select` returning `-1` with `errno` being set to `EBADF` (see the [man select](https://man7.org/linux/man-pages/man2/select.2.html))

**Comment by yhirose** [comment]
@tsilia, thanks for the suggestion. I updated the code.

**Review by tsilia — CHANGES_REQUESTED** [review]
Please see the comments in this review.
TLDR; the logic is incorrect now and I'd suggest to revert to the previously approved code.

File: httplib.h [hunk 1]
```diff
@@ -5723,13 +5721,26 @@ inline bool ClientImpl::send(Request &req, Response &res, Error &error) {
...
     auto is_alive = false;
     if (socket_.is_open()) {
-      is_alive = detail::select_write(socket_.sock, 0, 0) > 0;
+      // Check if the socket is still alive.
+      {
+        auto val = detail::select_read(socket_.sock, 0, 0);
+        if (val < 0) {
+          return false;
+        } else if (val == 0) {
+          is_alive = true;
```
[the remark this answers is not in the document]
Comment (yhirose) [reply httplib.h:5737]: Thanks for the suggestion. Somehow, I can't reproduce the problem "So now the client again always closes socket after each send"... But your suggestion sounds good to me, and I reverted and adjust is_socket_alive accordingly. I confirmed all the uni tests and your test (https://github.com/tsilia/poc-httplib-inf-keepalive) worked successfully. Please let me know if you find anything.

[the remark this answers is not in the document]
Comment (tsilia) [reply httplib.h]: > Somehow, I can't reproduce the problem "So now the client again always closes socket after each send"...

I'm totally positive that with that logic client always closed connection after each send:
1. There was a noticeable delay before each send (because client must reconnect again and again) which wouldn't come up with the correct logic (it's only noticeable using a 'real' network address, not `localhost`).
2. I've put some debug messages into the `detail::close_socket` function and around and these messages could be seen on every send:
```
./client 
Press enter key to stop client...
* Socket is not open!
Successfully posted 1882880898
!is_alive
close_socket
Successfully posted 1782988459
!is_alive
close_socket
Successfully posted 108701542
!is_alive
close_socket
Successfully posted 1110670616
!is_alive
close_socket
Successfully posted 1475065584

* Stopping client...
!is_alive
close_socket
Successfully posted 1284131091
close_socket
* Client stopped. Exiting...
```
3. You can see that that logic wasn't correct in terms of keeping the connection alive: the value of `is_alive` was first set to `true` when `val == 0` (`select` returned `0`), and then `is_alive` was unconditionally reassigned to the return value of `recv`.
Here's what was happening:
1. `select_read` returned 0 (socket had't been disconnected), `is_alive` was true.
2. `read_socket` returned 0 (again because no read/disconnect event), hence `is_alive` became false.

**Review by tsilia — APPROVED** [review]
It's working again, I double checked major cases.
Thanks!
